machinist/mnode/mnode.go

```diff
@@ -67,3 +76,91 @@ func (n *Node) BeginPolling() error {
 		}
 	}
 }
+
+// Todo(adam): perform rollbacks if enroll fails
+func (n *Node) Enroll() error {
+	if os.Geteuid() != 0 && n.config.RequireRoot {
+		return errors.New("this command must be run as root since it touches the /etc/ssh directory")
+	}
+	pubKey, privKey, err := kcerts.GenerateED25519()
+	if err != nil {
+		return err
+	}
+	hcr := &auth.HostCertificateRequest{
+		Hostcert: pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: ssh.MarshalAuthorizedKey(pubKey)}),
+		Hosts:    n.config.DnsNames,
+	}
+	resp, err := n.AuthClient.HostCertificate(context.Background(), hcr)
+	if err != nil {
+		fmt.Println("error here")
+		return err
+	}
+	if fName, exists := anyFileExist(
+		n.config.CaPublicKeyLocation,
+		n.config.HostKeyLocation, n.config.HostCertificate()); exists && !n.config.ReWriteConfigs {
+		return fmt.Errorf("cannot rewrite %s because it exists and rewriting is disabled", fName)
+	}
+
+	enrollConfig := n.config.enrollConfigs
+	// Pam Installer Steps
+	n.Log.Infof("Executing Pam installation steps")
+	if err := InstallLibPam(n.Log); err != nil {
+		return err
+	}
+	if err := InstallPamSSHDFile(enrollConfig.PamSSHDLocation, n.Log); err != nil {
+		return err
+	}
+	if err := InstallPamScript(enrollConfig.PamSecurityLocation, n.Log); err != nil {
+		return err
+	}
+
+	//// Nss AutoUser Setup
+	if err := InstallNssAutoUserConf(n.config.LibNssConfLocation, n.config.NssConfig()); err != nil {
+		return err
+	}
+	if err := InstallNssAutoUser(n.Log); err != nil {
+		return err
+	}
+
+	// SSHD installer steps
+	if err := os.MkdirAll(filepath.Dir(n.config.SSHDConfigurationLocation), os.ModePerm); err != nil {
+		return err
+	}
+	sshdConfigContent, err := ReadSSHDContent(n.config.CaPublicKeyLocation, n.config.HostKeyLocation, n.config.HostCertificate())
+	if err != nil {
+		return err
+	}
+	n.Log.Infof("Writing SSHD Configuration")
+	if err := ioutil.WriteFile(n.config.SSHDConfigurationLocation, sshdConfigContent, 0644); err != nil {
+		return err
+	}
+	n.Log.Infof("Writing CA Public Key Configuration")
+	if err := ioutil.WriteFile(n.config.CaPublicKeyLocation, resp.Capublickey, 0644); err != nil {
+		return err
+	}
+	n.Log.Infof("Writing Host Cert")
+	if err := ioutil.WriteFile(n.config.HostCertificate(), resp.Signedhostcert, 0644); err != nil {
+		return err
+	}
+	n.Log.Infof("Writing Host Key")
+	pemBytes, err := privKey.SSHPemEncode()
+	if err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(n.config.HostKeyLocation, pemBytes, 0644); err != nil {
+		return err
+	}
+	return nil
+}
+
+func anyFileExist(names ...string) (string, bool) {
+	for _, name := range names {
+		if _, err := os.Stat(name); err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
```

Comment:
minor issue: Stat could fail for any number of reasons. Eg, ssh is not installed, the parent directory does not exist, or is not writeable, ... If any of those failures happens, the code will output a message saying that the file already exists, very confusing.

I would personally use `lib/multierror` here, and change the function to ONLY return error. No error: none of the files exists. error: some of the file exists, or something else went sideways.

Basically, if `IsNotExist(err) true` -> no error added. Otherwise, error added. Should also simplify the logic, and show all errors in one go, which is more user friendly (eg, now, let's say two files exist, I remove one, run again, error again).